### PR TITLE
[Prototype] Add DependencyContext api

### PR DESCRIPTION
**Pull request by pakrym**
Issue: https://github.com/dotnet/cli/issues/376

@davidfowl  @anurse

### CODE REVIEW

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Extensions.DependencyModel.IO
```
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs]: Namespace with one class..

**Comment by pakrym** [comment]
Added writer

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs]: Why do we need an IO namespace?

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        public string Runtime { get; set; }
+
+        public IReadOnlyList<Library> CompileLibraries { get; }
+
+        public IReadOnlyList<Library> RuntimeLibraries { get; }
+
+
+        public static DependencyContext Load()
+        {
+            Assembly entryAssembly = null;
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: null?

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs]: Readers\Writers,

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextStrings.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+namespace Microsoft.Extensions.DependencyModel.IO
+{
+    public class DependencyContextStrings
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextStrings.cs]: internal

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextWriter.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Extensions.DependencyModel.IO
+{
+    public class DependencyContextWriter
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextWriter.cs]: Why does this library need a writer?

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs (hunk 2 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: Coreclr?

File: src/Microsoft.Extensions.DependencyModel/LibraryAssetDependency.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    public class Library
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/LibraryAssetDependency.cs]: struct

File: src/Microsoft.Extensions.DependencyModel/project.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/dotnet/cli"
+  },
+  "compilationOptions": {
+    "warningsAsErrors": true,
+    "emitEntryPoint": true
+  },
+  "dependencies": {
+    "Microsoft.DotNet.ProjectModel": "1.0.0-*",
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/project.json]: This library cannot depend on the project model.

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextWriter.cs (hunk 4 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextWriter.cs]: We can split it out somewhere else, but its nice to use same classes to generate deps file as to read it

File: src/Microsoft.Extensions.DependencyModel/project.json (hunk 6 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/project.json]: Again, we can put converter to another lib, i'm just not sure about witch one

File: src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs (hunk 1 above)
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/IO/DependencyContextReader.cs]: I'd just leave them in the main namespace.

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.ProjectModel.Compilation;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    public static class DependencyContextBuilder
+    {
+        public static  DependencyContext FromLibraryExporter(LibraryExporter libraryExporter, string target, string runtime)
+        {
+            var dependencies = libraryExporter.GetDependencies();
```
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: I'm pretty sure this leaves out the root project. I think we actually want the root project to be in the `DependencyContext`. We should use `GetAllExports` instead.

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs (hunk 2 above)
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: `Assembly.GetEntryAssembly` should be there now, and if not you can use reflection.

File: src/Microsoft.Extensions.DependencyModel/LibraryAssetDependency.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    public struct Library
```
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/LibraryAssetDependency.cs]: Filename doesn't match class name

**Comment by anurse** [comment]
Looks like a good start. The `deps` file format is still very much up in the air, but we can adjust that as we go.

**Comment by pakrym** [comment]
:up: :date:

**Comment by anurse** [comment]
Looks good! I'd like to hang on a bit before merging so we can try to land the specific file formats discussion first (to avoid confusion)

File: src/Microsoft.Extensions.DependencyModel/project.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    "warningsAsErrors": true,
+    "emitEntryPoint": true
+  },
+  "dependencies": {
+    "Newtonsoft.Json": "7.0.1"
+  },
+  "frameworks": {
+    "dnxcore50": {
+      "dependencies": {
+
```
Comment (troydai) [inline src/Microsoft.Extensions.DependencyModel/project.json]: why the white spaces?

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        public string Runtime { get; set; }
+
+        public IReadOnlyList<Library> CompileLibraries { get; }
+
+        public IReadOnlyList<Library> RuntimeLibraries { get; }
+
+
+        public static DependencyContext Load()
+        {
+            Assembly entryAssembly = (Assembly)typeof(Assembly).GetTypeInfo().GetDeclaredMethod("GetEntryAssembly").Invoke(null, null);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: var

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+            var path = Path.Combine(entryAssembly.Location, Path.GetFileNameWithoutExtension(entryAssembly.Location) + DepsExtension);
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: Read it from a resource

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+using System.Threading.Tasks;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.ProjectModel.Compilation;
+using Microsoft.DotNet.ProjectModel.Graph;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    public static class DependencyContextBuilder
+    {
+        public static  DependencyContext FromLibraryExporter(LibraryExporter libraryExporter, string target, string runtime)
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: extra space

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs (hunk 11 above)
Comment (anurse) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: Nothing embeds it yet. Let's not jump the gun and add a half-baked feature...

**Comment by pakrym** [comment]
:up: :date:

**Comment by pakrym** [comment]
:up: :date: 
I added actual compilation option and embedding code. I'd like to merge this and continue playing with mvc and other scenarios with library manager/exporter out of the way.

File: src/Microsoft.DotNet.Tools.Compiler/Program.cs [hunk 13]
```diff
@@ -319,16 +322,49 @@ private static bool CompileProject(ProjectContext context, string configuration,
...
+                var writer = new DependencyContextWriter();
+                var depsJsonFile = Path.Combine(intermediateOutputPath, context.ProjectFile.Name + "dotnet-compile.deps.json");
+                using (var fileStream = File.Create(depsJsonFile))
+                {
+                    writer.Write(dependencyContext, fileStream);
+                }
+
+                compilerArgs.Add($"--resource:\"{depsJsonFile}\",DependencyContext.json");
+
+                var refsFolder = Path.Combine(outputPath, "refs");
```
Comment (anurse) [inline src/Microsoft.DotNet.Tools.Compiler/Program.cs]: Create an issue for removing the `refs` folder from compiled output. We should be able to resolve them from the cache at runtime in this scenario. When publishing we will want to copy to a flat folder.

File: src/Microsoft.DotNet.Tools.Compiler/Program.cs [hunk 14]
```diff
@@ -319,16 +322,49 @@ private static bool CompileProject(ProjectContext context, string configuration,
...
+                    libraryExporter, context.TargetFramework.DotNetFrameworkName, context.RuntimeIdentifier);
+
+                var writer = new DependencyContextWriter();
+                var depsJsonFile = Path.Combine(intermediateOutputPath, context.ProjectFile.Name + "dotnet-compile.deps.json");
+                using (var fileStream = File.Create(depsJsonFile))
+                {
+                    writer.Write(dependencyContext, fileStream);
+                }
+
+                compilerArgs.Add($"--resource:\"{depsJsonFile}\",DependencyContext.json");
```
Comment (anurse) [inline src/Microsoft.DotNet.Tools.Compiler/Program.cs]: Rename the resource to `[assemblyname].deps.json`

**Comment by anurse** [comment]
:shipit: after the minor comments are resolved.

File: src/Microsoft.DotNet.Tools.Compiler/Program.cs (hunk 13 above)
Comment (pakrym) [inline src/Microsoft.DotNet.Tools.Compiler/Program.cs]: https://github.com/dotnet/cli/pull/425

**Pull request merged**